### PR TITLE
Implement transport client

### PR DESCRIPTION
**Pull request by rosstimothy**
Provides a client that can be used to connect to and interact with the transport service in `lib/srv/transport`. The client abstracts the fact that a gRPC stream is being used for the `net.Conn` created by `DialCluster` and `DialHost`.

This also moves `lib/utils/grpc/stream` to `api/utils/grpc/stream` so that the client and server can make use of the same stream abstractions.

Part of #19812

### CODE REVIEW

File: api/client/proxy/transport/transportv1/client.go [hunk 1]
```diff
@@ -0,0 +1,228 @@
...
+		return nil, nil, trace.Wrap(err)
+	}
+
+	// multiplex the frames to the correct handlers
+	var serveOnce sync.Once
+	go func() {
+		for {
+			req, err := stream.Recv()
+			if err != nil {
+				if errors.Is(err, io.EOF) {
```
Comment (strideynet) [inline api/client/proxy/transport/transportv1/client.go:158]: This err check treats an `io.EOF` the same as any other error - is there a missing behaviour here for an error that isn't an `io.EOF` ?

File: api/client/proxy/transport/transportv1/client_test.go [hunk 2]
```diff
@@ -0,0 +1,588 @@
...
+		s.Stop()
+	})
+
+	// Register service.
+	proxyv1.RegisterProxyServiceServer(s, srv)
+
+	// Start.
+	go func() {
+		if err := s.Serve(lis); err != nil {
+			panic(fmt.Sprintf("Serve returned err = %v", err))
```
Comment (strideynet) [inline api/client/proxy/transport/transportv1/client_test.go:515]: Is there an advantage to using a panic instead of `t.FailNow` or similar here ?

Comment (strideynet) [reply api/client/proxy/transport/transportv1/client_test.go:515]: I don't have particularly strong feelings (I'm not the world's biggest fan of `assert`/`require`) - but I do think that a large part of our test corpus currently uses `assert`/`require` and without any directive to suggest we are moving away from that, where it works, it's probably best to remain consistent with that.

File: api/client/proxy/transport/transportv1/client.go (hunk 1 above)
Comment (strideynet) [reply api/client/proxy/transport/transportv1/client.go:158]: Do we want to consider warn/debug logging for these ?

**Review by strideynet — APPROVED** [review]
Really nice and neat PR.

It definitely would be nice once all the parts of this come together for us to consider putting together some Go benchmarks E2E for this - it's definitely code I can see performance regressions occurring within.

**Review by gzdunek — APPROVED** [review]

**Pull request merged**